pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -344,3 +319,97 @@ def two_qubit_decomposition(U, wires):
+    match with the default decomposition of the single-qubit :class:`~.Rot` operations
+    as ``RZ RY RZ``. The most general form of the decomposition is:
+
+    .. figure:: ../../_static/two_qubit_decomposition.svg
+        :align: center
+        :width: 70%
+        :target: javascript:void(0);
+
+
+    where :math:`A, B, C, D` are :math:`SU(2)` gates.
+
+    .. note::
+
+        Currently, the decomposition is only implemented for two cases: when the
+        provided unitary can be expressed with no CNOTs (i.e., it is a tensor
+        product of two single-qubit operations), or when it can be expressed
+        using exactly 3, as in the graphic above. It generally works when the
+        provided unitary is sampled at random from :math:`U(4)`. The case of 1-
+        and 2-CNOTs will be implemented at a later time.
+
+    This decomposition can be applied automatically to all valid two-qubit
+    :class:`~.QubitUnitary` operations by applying the :func:`~pennylane.transforms.unitary_to_rot`
+    transform.
+
+    Args:
+        U (tensor): A 4 x 4 unitary matrix.
+        wires (Union[Wires, Sequence[int] or int]): The wires on which to apply the operation.
+
+    Returns:
+        list[Operation]: A list of operations that represent the decomposition
+        of the matrix U.
+
+    **Example**
+
+    Suppose we create a random element of :math:`U(4)`, and would like to decompose it
+    into elementary gates in our circuit.
+
+    >>> from scipy.stats import unitary_group
+    >>> U = unitary_group.rvs(4)
+    >>> U
+    array([[-0.29113625+0.56393527j,  0.39546712-0.14193837j,
+             0.04637428+0.01311566j, -0.62006741+0.18403743j],
+           [-0.45479211+0.25978444j, -0.52737418-0.5549423j ,
+            -0.23429057+0.10728103j,  0.16061807-0.21769762j],
+           [-0.4501231 +0.04065613j, -0.25558662+0.38209554j,
+            -0.04143479-0.56598134j,  0.12983673+0.49548507j],
+           [ 0.23899902+0.24800931j,  0.03374589-0.15784319j,
+             0.24898226-0.73975147j,  0.0269508 -0.49534518j]])
+
+    We can compute its decompositon like so:
+
+    >>> decomp = qml.transforms.two_qubit_decomposition(np.array(U), wires=[0, 1])
+    >>> decomp
+    [Rot(tensor(-1.69488788, requires_grad=True), tensor(1.06701916, requires_grad=True), tensor(0.41190893, requires_grad=True), wires=[0]),
+     Rot(tensor(1.57705621, requires_grad=True), tensor(2.42621204, requires_grad=True), tensor(2.57842249, requires_grad=True), wires=[1]),
+     CNOT(wires=[1, 0]),
+     RZ(0.4503059654281863, wires=[0]),
+     RY(-0.8872497960867665, wires=[1]),
+     CNOT(wires=[0, 1]),
+     RY(-1.6472464849278514, wires=[1]),
+     CNOT(wires=[1, 0]),
+     Rot(tensor(2.93239686, requires_grad=True), tensor(1.8725019, requires_grad=True), tensor(0.0418203, requires_grad=True), wires=[1]),
+     Rot(tensor(-3.78673588, requires_grad=True), tensor(2.03936812, requires_grad=True), tensor(-2.46956972, requires_grad=True), wires=[0])]
```

Comment:
Docstring is wonderful 💯 

Out of curiosity, if I am reading this correctly - the gates with parameters that are tensors are trainable and 'parametrize' the unitary, whereas the other gates (RZ, RY) are constants?

Reply:
Actually they all parametrize the unitary (12 from the tensors and 3 from here = 15 parameters needed for SU(4)). They are just obtained in different ways. Would it make sense for them to all be tensors?

Reply:
>Would it make sense for them to all be tensors?

I had a feeling they would just automatically _all_ be tensors, since they are derived from the original tensor unitary matrix 🤔 How come the middle three aren't?

Reply:
I would have thought so too I guess? This is how they are being computed:

```python
# The rotation angles can be computed as follows (any three eigenvalues can be used)
u = math.dot(Edag, math.dot(swap_U, E))
gammaU = math.dot(u, math.T(u))
evs, _ = math.linalg.eig(gammaU)

# To get consistent results in all interfaces, sort the eigenvalues first
angles = qml.math.sort([math.angle(ev) for ev in evs])
x, y, z = angles[0], angles[1], angles[2]

# Compute functions of the eigenvalues; there are different options in v1
# vs. v3 of the paper, I'm not entirely sure why. This is the version from v3.
alpha = (x + y) / 2
beta = (x + z) / 2
delta = (z + y) / 2
```

Maybe something to do with `eig`?

Reply:
Ohhh, maybe `qml.math.sort()` is seeing the input as a _list_, and so it just using vanilla numpy? I wonder if `math.angle(ev)` is a tensor, but `angles` is an array.